tests/contracts/map_corners_test.scilla

```diff
@@ -0,0 +1,569 @@
+(* test_query_empty_map_3 *)
+transition t15()
+  tname = "t15";
+
+  (* f_m3 should be [key1a][key2a] = [] *)
+  m3 <- f_m3;
+  m3_size = builtin size m3;
+  one = Uint32 1;
+  is_one = builtin eq m3_size one;
+  match is_one with
+  | True =>
+  | False =>
+    err = "Incorrect size of f_m3";
+    fail_msg tname err
+  end;
+
+  key1a = "key1a";
+  key2a = "key2a";
+  m2o = builtin get m3 key1a;
+  match m2o with
+  | Some m2 =>
+    m2_size = builtin size m2;
+    is_one = builtin eq m2_size one;
+    match is_one with
+    | True =>
+      m1o = builtin get m2 key2a;
+      match m1o with
+      | Some m1 =>
+        m1_size = builtin size m1;
+        zero = Uint32 0;
+        is_empty = builtin eq m1_size zero;
+        match is_empty with
+        | True =>
+        | False =>
+          err = "Incorrect size of f_m3[key1a][key2a]";
+          fail_msg tname err
+        end
+      | None =>
+        err = "Unexpected empty m1";
+        fail_msg tname err
+      end
+    | False =>
+      err = "Incorrect size of f_m3[key1]";
+      fail_msg tname err
+    end
+  | None =>
+    err = "Unexpected empty m2";
+    fail_msg tname err
+  end;
+
+  (* For the next test, insert a non-empty nested map into f_m2 *)
+  key1b = "key1b"; key2b = "key2b";
+  key1c = "key1c"; key2c = "key2c";
+  key1d = "key1d"; key2d = "key2d";
+  v1 = "420"; v2 = "421"; v3 = "422"; v4 = "423";
+  m2 = Emp String (Map String String);
+  m1 = Emp String String;
+  m2_full =
+    let m21 = builtin put m1 key2a v1 in
+    let m22 = builtin put m1 key2b v2 in
+    let m23 = builtin put m1 key2c v3 in
+    let m24 = builtin put m1 key2d v4 in
+    let m11 = builtin put m2 key1a m21 in
+    let m12 = builtin put m11 key1b m22 in
+    let m13 = builtin put m12 key1c m23 in
+    let m14 = builtin put m13 key1d m24 in
+    m14;
+  f_m2 := m2_full
+end
+
+(* test_query_update_fetch_nested *)
+transition t16()
+  tname = "t16";
+
+  (* Compare the entries *)
+  (* f_m2[key1a][key1a] = "420" *)
+  (* f_m2[key1b][key1b] = "421" *)
+  (* f_m2[key1c][key1c] = "422" *)
+  (* f_m2[key1d][key1d] = "423" *)
+  key1a = "key1a"; key2a = "key2a";
+  key1b = "key1b"; key2b = "key2b";
+  key1c = "key1c"; key2c = "key2c";
+  key1d = "key1d"; key2d = "key2d";
+
+  t1 <- f_m2[key1a][key2a];
+  t2 <- f_m2[key1b][key2b];
+  t3 <- f_m2[key1c][key2c];
+  t4 <- f_m2[key1d][key2d];
+  v1 = "420"; v2 = "421"; v3 = "422"; v4 = "423";
+
+  b1 = match t1 with
+      | Some t1v =>
+        builtin eq t1v v1
+      | None => False
+      end;
+
+  b2 = match t2 with
+      | Some t2v =>
+        builtin eq t2v v2
+      | None => False
+      end;
+
+  b3 = match t3 with
+      | Some t3v =>
+        builtin eq t3v v3
+      | None => False
+      end;
+
+  b4 = match t4 with
+      | Some t4v =>
+        builtin eq t4v v4
+      | None => False
+      end;
+
+  b =
+    let a1 = andb b1 b2 in
+    let a2 = andb b3 b4 in
+    andb a1 a2;
+
+  match b with
+  | True =>
+  | False =>
+    fail tname
+  end
```

Comment:
I think there is a test case missing (maybe it's there and I missed it, but it's important):

```
field triple_map : Map String (Map String (Map String String))
...
triple_map[key1][key2][key3] := "Hello World!";
delete triple_map[key1];
...
x <- triple_map[key1][key2][key3];
(* x should now be None *)
```

This is to ensure that deletion of a submap prevents lookup not only in that submap but also in any of its submaps.

Also, do we test the following:

```
map[key1][key2] := "Hello World!";
map[key1] := Emp String String;
x <- map[key1][key2];
```

I guess I am looking for tests where we delete or overwrite submaps, and then try to access entries in that submap. It seems like the sort of thing that could go wrong.

Reply:
Oh, we also need to test that we can write an entire map into a submap:

```
field map_field : Map String (Map String String)
...
map_val = Emp String String;
map_val = builtin put map_val key2 key3;
map_field[key1] := map_val;
x <- map_field[key1][key2]; (* Should be key3 *)
```

And also

```
field map_field : Map String (Map String String)
...
map_field[key1][key2] := "Hello World!";
map_val = Emp String String;
map_val = builtin put map_val key2 "!dlroW olleH";
map_field[key1] := map_val;
x <- map_field[key1][key2]; (* Should be "!dlroW olleH" *)
```

Reply:
>This is to ensure that deletion of a submap prevents lookup not only in that submap but also in any of its submaps.
This was _almost_ there. We do delete a submap, but I had forgotten to check that it is empty !. I have added this check now in `t7`. (This is already in the C++ unit tests, from which this entire contract was created, but when I was translating it, I missed this check : so a thank you :) )

>Also, do we test the following: (writing an empty map value)
This is already tested. For example in `t11`, `t12` and `t13`.

>Oh, we also need to test that we can write an entire map into a submap:
This is also already tested. At the end of `t5` we fill in `f_m2[key1b]` with a full sub-map and check for it in `t6`.